exporter/datadogexporter/factory.go

```diff
@@ -195,13 +238,20 @@ func (f *factory) createMetricsExporter(
 				if md.ResourceMetrics().Len() > 0 {
 					attrs = md.ResourceMetrics().At(0).Resource().Attributes()
 				}
-				go hostmetadata.RunPusher(ctx, set, newMetadataConfigfromConfig(cfg), hostProvider, attrs)
+				go hostmetadata.RunPusher(ctx, set, pcfg, hostProvider, attrs)
 			})
 
+			// Consume resources for host metadata
+			for i := 0; i < md.ResourceMetrics().Len(); i++ {
+				res := md.ResourceMetrics().At(i).Resource()
+				if err2 := metadataReporter.ConsumeResource(res); err2 != nil {
```

Comment:
These few lines are repeated a few times, any chance it can be refactored?

Reply:
Some of them are accessing `ResourceMetrics()` while others access `ResourceSpans()`. I guess we could use an interface, something like the following:

```go
type resourcerIterable interface {
  Len() int
  At(int) interface { Resource() pcommon.Resource }
}

func forEachResource(ri resourcerIterable, fn func()) error {
    // roughly the same code as above
}
```

I am not sure however what the performance impact would be. Do you think it is worth doing this? 

We could also just have three functions for this, one per signal. That would also reduce duplication but not by much

Reply:
I was thinking about something like:
```go
func consume(reporter *inframetadata.Reporter, res pcommon.Resource, logger *Logger) {
  if err := metadataReporter.ConsumeResource(res); err != nil {
    logger.Warn("failed to consume resource for host metadata", zap.Error(err), zap.Any("resource", res))
  }
}
```
Then 
```go
consume(reporter, md.ResourceMetrics().At(i).Resource(), set.Logger)
```
That way you don't need to the variable name `err` being shadowed

Reply:
Alright, I was trying to avoid duplication too much it seems :smile: done